### PR TITLE
feat(policies): add shared dtype management to policy base classes

### PR DESCRIPTION
**Pull request by HaomingSong**
## Summary / Motivation

Policy authors currently implement precision configuration and checkpoint dtype handling independently. This PR provides an extension interface in `PreTrainedConfig` and `PreTrainedPolicy`: a requested dtype, an observable runtime dtype, declarative FP32 exceptions, and a common initialization/loading flow.

## What changed

- Add `config.dtype`, accepting `float16`, `bfloat16`, `float32`, `float64`, or an unspecified value that resolves to PyTorch's default floating dtype.
- Add the read-only `policy.dtype`. It reports actual floating tensors.
- Add `post_init()` to cast parameters and buffers directly to their final precision, preserve shared parameters and existing gradients.
- Add `_fp32_modules` declaration on the policy class. `_fp32_tensor_names` resolves full module names of protected tensors.
- Add `from_pretrained(dtype=...)`. Explicit values override the supplied/saved config;
- Add the shared dtype resolver, extension documentation, and focused tests using a tiny policy.

This API uses materialized tensors and ordinary PyTorch casts. It introduces no quantization handling, meta-device initialization, `dtype_plan`, `torch_dtype` compatibility alias, or overrides of `.to()`.

## How was this tested (or how to run locally)

```bash
uv run pytest -q tests/policies/test_pretrained_dtype.py tests/configs tests/common
```

Coverage includes explicit and unspecified dtypes, global-default preservation, shared aliases and gradients, path boundaries, non-persistent buffers, CLI/config serialization, automatic checkpoint dtype detection, load overrides, and save/reload after an explicit cast.

Local CPU validation: **141 passed**. All pre-commit hooks for the staged foundation files passed. This is scoped API/config/checkpoint coverage, not a full run of every concrete policy.

## Checklist (required before merge)

- [x] Linting/formatting run on the changed files
- [x] Scoped tests pass locally
- [x] Documentation updated
- [ ] CI is green
- [ ] Community Review: I have reviewed another contributor's open PR and linked it here

## Reviewer notes

This feature uses materialized tensors and ordinary PyTorch casts. It introduces no quantization handling, meta-device initialization.

### CODE REVIEW

**Comment by pkooij** [comment]
@claude review, also look into any potential backward compatibility issues with existing checkpoints

**Comment by pkooij** [comment]
@claude this pr addresses policy specific changes, and is build on top of this pr, so if we merge them at same time would that take away your review points? https://github.com/huggingface/lerobot/pull/4596

**Review by pkooij — COMMENTED** [review]
I think second comment of Claude review makes sense, could we at least retarget 4596 to this pr, and make them stacked? and then address EO-1 and Lingbot flagged concerns. For the rest this pr looks good
